docarray/array/array/pushpull/__init__.py

```diff
@@ -0,0 +1,244 @@
+import logging
+from abc import abstractmethod
+from pathlib import Path
+from typing import TYPE_CHECKING, Dict, Iterable, Iterator, List, Optional, Type
+
+from typing_extensions import Protocol
+
+__cache_path__ = Path.home() / '.cache' / 'docarray-v2'
+
+if TYPE_CHECKING:  # pragma: no cover
+    from docarray import BaseDocument, DocumentArray
+
+
+class ConcurrentPushException(Exception):
+    """Exception raised when a concurrent push is detected."""
+
+    pass
+
+
+class PushPullLike(Protocol):
+    @staticmethod
+    def list(namespace: str, show_table: bool) -> List[str]:
+        ...
+
+    @staticmethod
+    def delete(name: str, missing_ok: bool) -> bool:
+        ...
+
+    @staticmethod
+    def push(
+        da: 'DocumentArray',
+        url: str,
+        public: bool,
+        show_progress: bool,
+        branding: Optional[Dict],
+    ) -> Dict:
+        ...
+
+    @staticmethod
+    def push_stream(
+        docs: Iterator['BaseDocument'],
+        url: str,
+        public: bool = True,
+        show_progress: bool = False,
+        branding: Optional[Dict] = None,
+    ) -> Dict:
+        ...
+
+    @staticmethod
+    def pull(
+        cls: Type['DocumentArray'],
+        name: str,
+        show_progress: bool,
+        local_cache: bool,
+    ) -> 'DocumentArray':
+        ...
+
+    @staticmethod
+    def pull_stream(
+        cls: Type['DocumentArray'],
+        name: str,
+        show_progress: bool,
+        local_cache: bool,
+    ) -> Iterator['BaseDocument']:
+        ...
+
+
+class PushPullMixin(Iterable['BaseDocument']):
+    """Mixin class for push/pull functionality."""
+
+    __backends__: Dict[str, PushPullLike] = {}
+    document_type: Type['BaseDocument']
+
+    @abstractmethod
+    def __len__(self) -> int:
+        ...
+
+    @classmethod
+    def get_backend(cls, protocol: str) -> PushPullLike:
```

Comment:
Maybe we should use Literal here ? https://docs.python.org/3/library/typing.html#typing.Literal


```suggestion
    def get_backend(cls, protocol: Literal['jnaai', 's3', 'file']) -> PushPullLike:
```